### PR TITLE
[gym] Provide bcsymbolmap path to dsymutil for each architecture

### PR DESCRIPTION
**Pull request by leo150**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Currently `gym` doesn't map bcsymbolmap files with debug symbols for the project frameworks. The problem appears in the log as a warning: 

> [09:44:25]: Mapping dSYM(s) using generated BCSymbolMaps
> [09:44:25]: dsymutil --symbol-map /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.1.0-2\ 2019-08-29\ 09.41.05.xcarchive/BCSymbolMaps /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.0.0-1\ 2019-08-29\ 09.41.05.xcarchive/dSYMs/Alamofire.framework.dSYM
> [09:44:25]: warning: /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.0.0-1 2019-08-29 09.41.05.xcarchive/BCSymbolMaps/Alamofire-armv7.bcsymbolmap: No such file or directory. Not unobfuscating.
> [09:44:25]: warning: /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.0.0-1 2019-08-29 09.41.05.xcarchive/BCSymbolMaps/Alamofire-arm64.bcsymbolmap: No such file or directory. Not unobfuscating.

It can be observed in the logs of at least [16 issues](https://github.com/fastlane/fastlane/issues?utf8=%E2%9C%93&q=No+such+file+or+directory.+Not+unobfuscating). 
These two address the issue directly: #10902 and #14197

As result, dSYM files don't unobfuscate stacktrace completely:

> #0. Crashed: com.apple.main-thread
> 0  libswiftCore.dylib     0x21dee4ce0 specialized \_assertionFailure(_:_:file) + 260
> 1  libswiftCore.dylib      0x21bb071bc \_assertionFailure(_:_:file) + 42
> 2  Alamofire                   0x104e17414 (Missing) 
> 3  MyApp                       0x104471724 TestViewController.simulateCrash() (\<compiler-generated\>)
> ...

### Description
Proposed solution executes `dwarfdump` for each dSYM file to extract UUID of each architecture. For example: 
```
$ dwarfdump --uuid ./dSYMs/Alamofire.framework.dSYM
UUID: 1B8FF487-DC57-3F37-11DF-D466877AECF9 (armv7) ...
UUID: FFA4DD5D-741C-3ECF-34FB-16971D545948 (arm64) ...
```

Bcsymbolmap files are stored in BCSymbolMaps folder in the format `\<architecture UUID\>.bcsymbolmap`. So the goal is to map UUIDs from dSYM file into filenames of each bcsymbolmap.

Then `dsymutil` can link the DWARF debug information from correct path:

> [09:44:25]: Mapping dSYM(s) using generated BCSymbolMaps
> [09:44:25]: $ dsymutil --symbol-map /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.0.0-1\ 2019-08-29\ 09.41.05.xcarchive/BCSymbolMaps/1B8FF487-DC57-3F37-11DF-D466877AECF9.bcsymbolmap /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.0.0-1\ 2019-08-29\ 09.41.05.xcarchive/dSYMs/Alamofire.framework.dSYM
> [09:44:25]: $ dsymutil --symbol-map /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.0.0-1\ 2019-08-29\ 09.41.05.xcarchive/BCSymbolMaps/FFA4DD5D-741C-3ECF-34FB-16971D545948.bcsymbolmap /Users/user/Library/Developer/Xcode/Archives/2019-08-29/MyApp-1.0.0-1\ 2019-08-29\ 09.41.05.xcarchive/dSYMs/Alamofire.framework.dSYM
> ...

### Testing Steps

After rebuilding with `gym` and uploading corrected dSYMs to Crashlytics, the issue is resolved:

> #0. Crashed: com.apple.main-thread
> 0  libswiftCore.dylib     0x21dee4ce0 specialized \_assertionFailure(_:_:file) + 260
> 1  libswiftCore.dylib      0x21bb071bc \_assertionFailure(_:_:file) + 42
> 2  Alamofire                   0x1f01d64c4 TestExceptionClass.produceCrash() + 31 (AFError.swift:31)
> 3  MyApp                       0x104378824 TestViewController.simulateCrash()  + 42 (TestViewController.swift:42)
> ...

### CODE REVIEW

**Comment by leo150** [comment]
@googlebot I signed it!

**Comment by janpio** [comment]
Are there cases where `dwarfdump --uuid ...` might not return the expected values and it would be better to bring back the old, previous behavior? 

Could this be a breaking change for someone?

File: gym/lib/gym/runner.rb [hunk 1]
```diff
@@ -150,11 +150,20 @@ def compress_and_move_dsym
...
+          dwarfdump_result = Helper.backticks(dwarfdump_command.join(" "), print: false)
+          architecture_uuids = dwarfdump_result.split("\n").map { |info| info.split(" ")[1] }
+
+          architecture_uuids.each do |uuid|
+            command = []
+            command << "dsymutil"
+            command << "--symbol-map #{bcsymbolmaps_directory.shellescape}/#{uuid}.bcsymbolmap"
+            command << dsym.shellescape
+            Helper.backticks(command.join(" "), print: !Gym.config[:silent])
+          end
```
Comment (joshdholtz) [inline gym/lib/gym/runner.rb]: 1. Can we add a comment on why we are splitting on `[1]`? (see below)
2. Can we also do something like below where we map the `architecture_uuids` to an array of paths and then iterate over the paths? This way we can easily default back to original implementation if there are no uuids.

```rb
          dwarfdump_command = []
          dwarfdump_command << "dwarfdump"
          dwarfdump_command << "--uuid #{dsym.shellescape}"

          # Gets UUIDs for all architectures using dwarddump
          dwarfdump_result = Helper.backticks(dwarfdump_command.join(" "), print: false)
          architecture_uuids = dwarfdump_result.split("\n").map { |info| info.split(" ")[1] }

          symbol_map_paths = architecture_uuids.map do |uuid|
            "#{bcsymbolmaps_directory.shellescape}/#{uuid}.bcsymbolmap"
          end

          symbol_map_paths << bcsymbolmaps_directory.shellescape if symbol_map_paths.empty?

          symbol_map_paths.each do |path|
            command = []
            command << "dsymutil"
            command << "--symbol-map #{path}"
            command << dsym.shellescape
            Helper.backticks(command.join(" "), print: !Gym.config[:silent])
          end
```

Thoughts? 😊

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Thanks for looking into this! I can't say I would have found this solution 🙃 I have on request about that I think would make this a bit safer feeling and have the previous implementation as a backup (if needed)

File: gym/lib/gym/runner.rb (hunk 1 above)
Comment (leo150) [reply gym/lib/gym/runner.rb]: Thank you for contributing! It is definitely better to add some additional checks.

If we expect `dwarddump` to not return UUIDs in some cases or return an invalid text, we might add a verification for `dwarddump` result. Regex is based on [this](https://developer.apple.com/library/archive/technotes/tn2151/_index.html) technical note:

> Convert the UUID of the binary image to a 32 character string seperated in groups of 8-4-4-4-12 (XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX). Note that all letters must be uppercased.

AppStore dSYM files have lowercased names, so we could be not so strict as well and allow lowercased symbols. As result: 

```ruby
uuid_regex = /[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}/

if Dir.exist?(bcsymbolmaps_directory)
  UI.message("Mapping dSYM(s) using generated BCSymbolMaps")
  available_dsyms.each do |dsym|
    dwarfdump_command = []
    dwarfdump_command << "dwarfdump"
    dwarfdump_command << "--uuid #{dsym.shellescape}"

    dwarfdump_result = Helper.backticks(dwarfdump_command.join(" "), print: false)
    architecture_infos = dwarfdump_result.split("\n")
    architecture_uuids = architecture_infos.map { |info|
      info_array = info.split(" ")
      uuid = info_array[1]

      if uuid.nil? || !uuid.match(uuid_regex)
        nil
      else
        uuid
      end
    }.reject{ |uuid| uuid.nil? }

    symbol_map_paths = architecture_uuids.map do |uuid|
      "#{bcsymbolmaps_directory.shellescape}/#{uuid}.bcsymbolmap"
    end

    symbol_map_paths << bcsymbolmaps_directory.shellescape if symbol_map_paths.empty?

    symbol_map_paths.each do |path|
      command = []
      command << "dsymutil"
      command << "--symbol-map #{path}"
      command << dsym.shellescape
      Helper.backticks(command.join(" "), print: !Gym.config[:silent])
    end
  end
end
```

Comment (leo150) [reply gym/lib/gym/runner.rb]: So I've tested and pushed this combined code above with additional `dwarfdump` output validation and a fallback to the original implementation. Also applied `rubocop` corrections. 

I'd avoid adding any comments, just instead to make the variable names more explicit and readable. For example, splitting on `[1]` is `uuid = info_array[1]` now. Please reopen the conversation if any additional changes are required.

**Review by joshdholtz — APPROVED** [review]
This looks 💯 Thanks for making those changes! This will be out in the next release ❤️

**Pull request merged**